### PR TITLE
Add Deformable Convolution operation.

### PR DESCRIPTION
**Pull request by pedrofreire**
For #1416, this adds the deformable convolution operation, as described in Deformable Convolutional Networks (https://arxiv.org/abs/1703.06211).

- The code is based on https://github.com/open-mmlab/mmdetection/blob/master/mmdet/ops/dcn/src/deform_conv_cuda.cpp ; the whole code was modified and refactored to remove redundancies and increase clarity, and to adapt it to torchvision.

- The CPU part is a direct copy of the CUDA code; it might make sense to do follow-up adjustments in the CPU code to simplify it / optimize it, or to reuse functionality between CPU and CUDA..

- We also add tests; they likely can be made more robust by doing multiple calls with random parameters.

### CODE REVIEW

**Comment by pedrofreire** [comment]
I did not found which linter is used for C++, so the C++/CUDA code is currently without linting.
I also have to check if the code is Python 2 compatible :)

File: torchvision/ops/deform_conv.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
+from typing import Tuple
+
+import torch
+from torch import nn, Tensor
+from torch.nn.modules.utils import _pair
+from torch.jit.annotations import List
+
+
+def deform_conv(input, offset, weight, stride=(1, 1), pad=(0, 0), dilation=(1, 1), n_parallel_imgs=64):
```
Comment (fmassa) [inline torchvision/ops/deform_conv.py]: Can you name it `deform_conv2d`, because it's a conv2d operator?

File: torchvision/ops/deform_conv.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+    Arguments:
+        input (Tensor[batch_sz, in_channels, in_h, in_w]): input tensor
+        offset (Tensor[batch_sz, 2 * n_offset_grps * weight_h * weight_w, out_h, out_w])
+        weight (Tensor[out_channels, in_channels // n_weight_grps, weight_h, weight_w]):
+            convolution weights, with n_weight_grps different connection groups
+        stride (int or Tuple[int, int]): distance between convolution centers
+        pad (int or Tuple[int, int]): height/width of padding of zeroes around each image
+        dilation (int or Tuple[int, int]): point distance in convolution grid
+        n_parallel_imgs (int): Number of images to be processed at once; does not change
```
Comment (fmassa) [inline torchvision/ops/deform_conv.py]: I'm not sure this is something that should be exposed to the user.
In PyTorch, we try to limit the number of knobs that we expose to the users and we try to use good defaults instead.

While I understand the need for this (if we just make by default `n_parallel_imgs = batch_size` the amount of memory used will be pretty large), I think we should either let it 1 or `batch_size` by default, or just make a reasonable guess.

The reason why I'm saying this is because in PyTorch, `nn.functional.conv2d` doesn't expose such options, and I think we should follow accordingly.

File: torchvision/ops/deform_conv.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+            behavior, only used for performance purposes
+
+    Returns:
+        output (Tensor[batch_sz, out_channels, out_h, out_w]): result of convolution
+    """
+
+    stride_h, stride_w = stride
```
Comment (fmassa) [inline torchvision/ops/deform_conv.py]: Can you use `_pair` for `stride`, `padding` and `dilation`?

File: torchvision/ops/deform_conv.py (hunk 1 above)
Comment (fmassa) [reply torchvision/ops/deform_conv.py]: Also, can you rename `pad` to `padding` to make it consistent with pytorch `conv2d`, and change the order to be `input, weight, offset, ...`?

File: torchvision/ops/deform_conv.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    _, n_in_channels, in_h, in_w = input.shape
+
+    n_offset_grps = offset.shape[1] // (2 * weights_h * weights_w)
+    n_weight_grps = n_in_channels // weight.shape[1]
+
+    return torch.ops.torchvision.deform_conv(
+        input,
+        offset,
+        weight,
+        *stride,
```
Comment (fmassa) [inline torchvision/ops/deform_conv.py]: This is not Python2 compatible I think, and we are for now keeping Python2 compatibility until Jan 2020.

File: torchvision/ops/deform_conv.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    """
+    def __init__(self, stride=1, pad=0, dilation=1, n_parallel_imgs=64):
+        super(DeformConv, self).__init__()
+        self.stride = _pair(stride)
+        self.pad = _pair(pad)
+        self.dilation = _pair(dilation)
+        self.n_parallel_imgs = n_parallel_imgs
+
+    def forward(self, input, offset, weight):
+        return deform_conv(input, offset, weight, stride=self.stride, pad=self.pad,
```
Comment (fmassa) [inline torchvision/ops/deform_conv.py]: One of the benefits from the `nn.Module` interface of PyTorch is that it manages the trainable weights / buffers for us.

In this case, the trainable weights are `weight` and `offset`. Can you do something similar, following [the implementation of `nn.Conv2d`](https://github.com/pytorch/pytorch/blob/7782f4bc50f5231922509e7c929a225d55bb170d/torch/nn/modules/conv.py#L40-L41) (but using `torch.empty` instead of `torch.Tensor`)? Also, please don't forget to add some weight initialization, for example [for example the one from PyTorch](https://github.com/pytorch/pytorch/blob/7782f4bc50f5231922509e7c929a225d55bb170d/torch/nn/modules/conv.py#L48-L53) (or another one that is better adapted for DeformConv

File: torchvision/ops/deform_conv.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        weight,
+        *stride,
+        *pad,
+        *dilation,
+        n_weight_grps,
+        n_offset_grps,
+        n_parallel_imgs)
+
+
+class DeformConv(nn.Module):
```
Comment (fmassa) [inline torchvision/ops/deform_conv.py]: Can you rename this to `DeformConv2d`?

File: torchvision/ops/deform_conv.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    """
+    See deform_conv
+    """
+    def __init__(self, stride=1, pad=0, dilation=1, n_parallel_imgs=64):
```
Comment (fmassa) [inline torchvision/ops/deform_conv.py]: I think the API should be similar to that of `nn.Conv2d`, exposing `num_input_planes` etc, and creating the `weight` tensor internally.

File: torchvision/ops/deform_conv.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+        self.dilation = _pair(dilation)
+        self.n_parallel_imgs = n_parallel_imgs
+
+    def forward(self, input, offset, weight):
+        return deform_conv(input, offset, weight, stride=self.stride, pad=self.pad,
+                           dilation=self.dilation, n_parallel_imgs=self.n_parallel_imgs)
+
+    def __repr__(self):
+        tmpstr = self.__class__.__name__ + '('
+        tmpstr += 'output_size=' + str(self.output_size)
```
Comment (fmassa) [inline torchvision/ops/deform_conv.py]: This `__repr__` should be changed I think? ;-)

File: torchvision/ops/deform_conv.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
+from typing import Tuple
```
Comment (fmassa) [inline torchvision/ops/deform_conv.py]: Can you import `Tuple` from `from torch.jit.annotations import Tuple`? So that if the user doesn't have `typing` installed, it will still work.

File: test/test_ops.py [hunk 10]
```diff
@@ -1,15 +1,18 @@
 from __future__ import division
+import math
+from typing import Tuple
```
Comment (fmassa) [inline test/test_ops.py]: Can you import from `torch.jit.annotations`, instead of `typing`? Same reason as below

File: torchvision/ops/deform_conv.py [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+        self.kernel_size = _pair(kernel_size)
+        self.stride = _pair(stride)
+        self.padding = _pair(padding)
+        self.dilation = _pair(dilation)
+        self.groups = groups
+        self.offset_groups = offset_groups
+
+        self.weight = Parameter(torch.empty(out_channels, in_channels // groups, kernel_size[0], kernel_size[1]))
+
+        self.offset_conv = nn.Conv2d(
```
Comment (fmassa) [inline torchvision/ops/deform_conv.py]: After looking at it again, I think that it might be better to let the `offset` to be an additional input in the model forward.
Indeed, the user can just want to have a set of convolutions (potentially with residuals) to compute the offsets, and if we force it here to be a specific one this is a bit limiting.

File: torchvision/ops/deform_conv.py [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+        init.kaiming_uniform_(self.weight, a=math.sqrt(5))
+        init.zeros_(self.offset_conv.weight)
+        init.zeros_(self.offset_conv.bias)
+        if self.bias is not None:
+            fan_in, _ = init._calculate_fan_in_and_fan_out(self.weight)
+            bound = 1 / math.sqrt(fan_in)
+            init.uniform_(self.bias, -bound, bound)
+
+    def forward(self, input):
+        offset = self.offset_conv.to(device=input.device, dtype=input.dtype)(input)
```
Comment (fmassa) [inline torchvision/ops/deform_conv.py]: We let to the user to convert the model to the desired device, and we don't perform automatic casting inside the forward.
This is safer because it might avoid the user to perform (implicitly) numerous memory copies from CPU to GPU.

File: torchvision/ops/deform_conv.py [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+            each image. Default: 0
+        dilation (int or Tuple[int, int]): the spacing between kernel elements. Default: 1
+
+    Returns:
+        output (Tensor[batch_sz, out_channels, out_h, out_w]): result of convolution
+    """
+
+    out_channels = weight.shape[0]
+    if bias is None:
+        bias = torch.zeros(out_channels, device=input.device, dtype=input.dtype)
```
Comment (fmassa) [inline torchvision/ops/deform_conv.py]: We can leave this as is, but ideally if the `bias` is None we would want to avoid adding it in the code.

One easy way of doing it would be something like
```python
out = # output of deform_conv2d
if bias is not None:
    out = out + bias[:, None, :, :]
```
this is slightly less efficient than if we had it in the kernel, but saves some compute if the bias doesn't exist.
Also note that we can handle this on the C++ level with `c10::optional`.

File: torchvision/csrc/cpu/DeformConv_cpu.cpp [hunk 14]
```diff
@@ -0,0 +1,985 @@
...
+    const at::Tensor& offset,
+    const at::Tensor& bias,
+    std::pair<int, int> stride,
+    std::pair<int, int> pad,
+    std::pair<int, int> dilation,
+    int n_weight_grps,
+    int n_offset_grps) {
+  const int batch_sz = input.size(0);
+  const int n_parallel_imgs =
+      get_greatest_divisor_below_bound(batch_sz, kMaxParallelImgs);
```
Comment (fmassa) [inline torchvision/csrc/cpu/DeformConv_cpu.cpp]: nice!

File: torchvision/ops/deform_conv.py (hunk 5 above)
Comment (fmassa) [reply torchvision/ops/deform_conv.py]: my bad, the `offset` is not trainable

File: torchvision/ops/deform_conv.py (hunk 11 above)
Comment (pedrofreire) [reply torchvision/ops/deform_conv.py]: Yeah, I was unsure on that. It does make creating the Module more complex (e.g. you have to compute the output size to create the offset), but it is indeed limiting to force the training of the offset.

File: torchvision/ops/deform_conv.py (hunk 13 above)
Comment (pedrofreire) [reply torchvision/ops/deform_conv.py]: I guess we can do this as a follow-up.
Btw, what is the best way to handle optionals and autograd::Variable?

Comment (fmassa) [reply torchvision/ops/deform_conv.py]: > I guess we can do this as a follow-up.

Sure, this is not a blocker for this PR

> Btw, what is the best way to handle optionals and autograd::Variable?

I would have expected that `c10::optional` would be enough, but if that's not the case I'd need to look a bit more. Let's let this in a follow-up

File: torchvision/ops/deform_conv.py [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+
+import torch
+from torch import nn, Tensor
+from torch.nn import init
+from torch.nn.parameter import Parameter
+from torch.nn.modules.utils import _pair
+from torch.jit.annotations import Optional, Tuple
+
+
+def deform_conv2d(input, weight, offset, bias=None, stride=(1, 1), padding=(0, 0), dilation=(1, 1)):
```
Comment (fmassa) [inline torchvision/ops/deform_conv.py]: nit: Given that `offset` can be seen as another input, I think it might make sense to move it so that the order of the arguments is `input, offset, weight, ...`
No need to change the order of the arguments in the C++ extension, just in the Python-facing function

**Comment by breznak** [comment]
Thanks for adding DCN! :100: 

I have a question, does this support `DCNv2`? 
https://github.com/CharlesShang/DCNv2 
https://github.com/chengdazhi/Deformable-Convolution-V2-PyTorch 

Also what is the difference between the two repos?

**Comment by pedrofreire** [comment]
This PR does not include DCNv2 unfortunately, though it should be straightforward, using this PR as a reference, to port them from mmdetection or one of the repos you mentioned (I have not looked for the differences between them, if there are any).

**Comment by breznak** [comment]
Thank you for the swift answer! I'm not sure if I'm upto the task, but I might try. Or someone can join me.

**Comment by jinfagang** [comment]
Any updates on this issue?

**Comment by fmassa** [comment]
@jinfagang which updates are you talking about?

**Comment by pedrofreire** [comment]
I guess he is talking about having support for DCNv2 (this PR only included the original DCN) - since that is what the most recent comments discuss about.

**Comment by fmassa** [comment]
Thanks for the clarification @pedrofreire 
If that's indeed the question, then no updates for DCNv2 for now.

**Comment by idow09** [comment]
Hi! Thanks for the useful op.
Why can't I find this anywhere on the official documentation?

**Comment by fmassa** [comment]
@idow09 we forgot to add it (plus PS RoIAlign / PS RoI Pool) to https://github.com/pytorch/vision/blob/master/docs/source/ops.rst

A PR adding those is welcome.

**Comment by idow09** [comment]
Okay I'll see if I can manage. Thanks again
